filters/include/pcl/filters/impl/crop_hull.hpp

```diff
@@ -113,72 +92,83 @@ pcl::CropHull<PointT>::getHullCloudRange ()
       if (pt[i] > cloud_max[i]) cloud_max[i] = pt[i];
     }
   }
-  
+
   return (cloud_max - cloud_min);
 }
 
 //////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 template<typename PointT> template<unsigned PlaneDim1, unsigned PlaneDim2> void 
-pcl::CropHull<PointT>::applyFilter2D (PointCloud &output)
+pcl::CropHull<PointT>::applyFilter2D (std::vector<int> &indices)
 {
-  for (std::size_t index = 0; index < indices_->size (); index++)
+  Eigen::Vector4f const minPt = crop_box_.getMin();
+  Eigen::Vector4f const maxPt = crop_box_.getMax();
+  std::vector<bool> cropInlierMask(input_->size(), true);
+  for (const index_t idx : *indices_)
   {
-    // iterate over polygons faster than points because we expect this data
-    // to be, in general, more cache-local - the point cloud might be huge
-    std::size_t poly;
-    for (poly = 0; poly < hull_polygons_.size (); poly++)
+    Eigen::Vector3f pt = input_->points[idx].getVector3fMap();
+    if (pt[PlaneDim1] < minPt[PlaneDim1] || pt[PlaneDim1] > maxPt[PlaneDim2] ||
+        pt[PlaneDim2] < minPt[PlaneDim2] || pt[PlaneDim2] > maxPt[PlaneDim2])
     {
-      if (isPointIn2DPolyWithVertIndices<PlaneDim1,PlaneDim2> (
-              input_->points[(*indices_)[index]], hull_polygons_[poly], *hull_cloud_
-         ))
-      {
-        if (crop_outside_)
-          output.push_back (input_->points[(*indices_)[index]]);
-        // once a point has tested +ve for being inside one polygon, we can
-        // stop checking the others:
-        break;
+      cropInlierMask.at(idx) = false;
+      if (!crop_outside_) {
+        indices.push_back(idx);
+      }
+      else {
+        removed_indices_->push_back(idx);
```

Comment:
This logic should be a lot more complicated depending on `extract_removed_indices_` and `negative_` and `crop_outside_`. Right now, we're breaking contract (which can be remedied by making `setRemovedIndices` as private and `extract_removed_indices_` to true)

If we can resolve the previous conversation, it would be as simple as it is now without breaking contract.

Else, I'd recommend using pointers to refer to `inlier_indices` and `outlier_indices` and using `extract_removed_indices_` to conditionally fill the latter.

Reply:
Should I introduce new public method `setRemovedIndices`?
`extract_removed_indices` is a member of class `Filter`. If introduce setter, it must be in `CropHull` or in `Filter`, which is base class for `CropHull`?

Reply:
changed API: add flag `extract_removed_indices` to `CropHull` constructor [default=false]